### PR TITLE
resync libretro port with upstream trunk

### PR DESCRIPTION
**Pull request by rtretiakov**
This pull request syncs everything in libretro up with the trunk. Ignore the makefiles, all of that stuff is specific to libretro for now. We're going to change some of the things discussed in the previous pull request in the future.

### CODE REVIEW

File: bsnes/GNUmakefile [hunk 1]
```diff
@@ -62,5 +86,6 @@ include $(ui)/GNUmakefile
 -include obj/*.d
 
 clean:
+	rm -f obj/*
```
Comment (orbea) [inline bsnes/GNUmakefile]: Shouldn't the `delete` do this, looks like a bad hack...

File: bsnes/target-libretro/GNUmakefile [hunk 2]
```diff
@@ -8,9 +12,19 @@ obj/libretro.o: target-libretro/libretro.cpp
 
 all: $(objects)
 ifeq ($(platform),linux)
-	$(strip $(compiler) -o out/bsnes_libretro.so -shared $(objects) -Wl,--no-undefined -Wl,--version-script=target-libretro/link.T -Wl,-Bdynamic -lpthread -ldl -lgomp)
+	$(strip $(compiler) -o out/bsnes_hd_libretro.so -shared $(objects) -Wl,--no-undefined -Wl,--version-script=target-libretro/link.T -Wl,-Bdynamic -lpthread -ldl -lgomp)
```
Comment (orbea) [inline bsnes/target-libretro/GNUmakefile]: Please keep the `bsnes_libretro` name and fix the name in libretro instead, there are 0 conflicting cores.

File: bsnes/GNUmakefile [hunk 3]
```diff
@@ -14,12 +18,26 @@ endif
 
 # binaries built with this flag are faster, but are not portable to multiple machines.
 ifeq ($(local),true)
-  flags += -march=native
+  ifneq ($(platform),$(filter $(platform),macos ios-arm64 tvos-arm64))
+    flags += -march=native
```
Comment (orbea) [inline bsnes/GNUmakefile]: `local` should be set to `false` for libretro probably instead of adding platform hacks.

**Comment by byuu** [comment]
Hopefully I can work through this tomorrow ... been a bit busy clearing the issues first.

**Comment by byuu** [comment]
Sorry for the delay.

ifeq ($(platform),$(filter $(platform),macos ios-arm64 tvos-arm64))

I'm opposed to doing this for macOS. OpenMP definitely works on macOS with the latest toolchain, and it's essential for HD mode 7 to scale up high. If it really doesn't work on the arm64 targets, I'm willing to add this code for those two targets, but I don't like it ...

ifneq ($(platform),$(filter $(platform),macos ios-arm64 tvos-arm64))
  flags += -march=native

The binary should be built with local=false for this. I added it to the target-libretro/GNUmakefile for now. I guess users can opt-in to it if they compile it themselves.

ifeq ($(platform),win)
  override platform = windows
else ifeq ($(platform),mingw)
  override platform = windows
else ifeq ($(platform),osx)
  override platform = macos
else ifeq ($(platform),unix)
  override platform = linux
else ifeq ($(platform),x)
  override platform = linux
endif

I don't understand this. Why are you doing this? Can you put it into target-libretro/GNUmakefile? I'm much less pick about things added there.

else ifeq ($(platform), ios-arm64)
  flags += -fPIC -miphoneos-version-min=11.0 -Wno-error=implicit-function-declaration -DHAVE_POSIX_MEMALIGN
  options += -dynamiclib
else ifeq ($(platform), tvos-arm64)
   flags += -fPIC -mtvos-version-min=11.0 -Wno-error=implicit-function-declaration -DHAVE_POSIX_MEMALIGN
   options += -dynamiclib

Is this possible to put into target-libretro/GNUmakefile? I don't support those targets in any of my applications.

	rm -f obj/*

Should be redundant. Is $(call delete,obj/*) not working? We should fix that if so.

ifneq ($(platform),$(filter $(platform),macos ios-arm64 tvos-arm64))
  flags += -fopenmp
endif

Oh, we should be using "openmp := true/false" here anyway.

I'd like to ask again if we can figure out OpenMP here because it's really, really essential for HD mode 7. If we really can't do it ... I'll add the disabling you requested.

	$(strip $(compiler) -o out/bsnes_hd_libretro.dylib -shared $(objects) -lpthread -ldl)
else ifeq ($(platform), ios-arm64)
    ifeq ($(IOSSDK),)
       IOSSDK := $(shell xcodebuild -version -sdk iphoneos Path)
    endif
	$(strip c++ -arch arm64 -marm -miphoneos-version-min=11.0 -isysroot $(IOSSDK) -o out/bsnes_hd_libretro_ios.dylib -shared $(objects) -lpthread -ldl)
else ifeq ($(platform), tvos-arm64)
    ifeq ($(IOSSDK),)
       IOSSDK := $(shell xcodebuild -version -sdk appletvos Path)
    endif
	$(strip c++ -arch arm64 -marm -mtvos-version-min=11.0 -isysroot $(IOSSDK) -o out/bsnes_hd_libretro_tvos.dylib -shared $(objects) -lpthread -ldl)

I won't call it bsnes_hd, but I merged the rest of this.

	//fixes an errant scanline on the title screen due to writing to PPU registers too late
	if(title == "FIREPOWER 2000") emulator->configure("Hacks/PPU/RenderCycle", 32);

Done.
		auto base = Location::base(base_name.transform("\\", "/"));

Done.

(resources.hpp)

Done. Maybe we should work sourcery into the build process, heh. Maybe not worth it though.

**Comment by rtretiakov** [comment]
Thanks for taking care of the merge.

Yes, all of the makefile stuff is basically hacks that are not intended to go upstream. All of that platform override stuff was done to make it build on the current buildbot which is going to be changed soon, and hopefully we can get rid of those hacks when the new one is running.
I think we can safely start calling it simply bsnes on our side, as @orbea pointed out here, there does not seem to actually be a conflict with that core name in the rest of the libretro ecosystem. That supposed conflict was the entire reason for not calling it bsnes in the first place.

The biggest impediment to being fully in sync right now is the builds. I think I can focus on trying to get that in order now that things are moving along.

**Comment by byuu** [comment]
That sounds awesome, thank you for all your hard work!

**Pull request closed without merging**